stablehlo/dialect/Base.cpp

```diff
@@ -300,6 +300,13 @@ FailureOr<TensorType> inferTypeWithCustomFn(
         Optional<Location>, int64_t, int64_t, int64_t, int64_t, int64_t)>
         inferDimAndBoundFn) {
   auto rank = rankedTypes[0].getRank();
+  for (auto& type : rankedTypes) {
```

Comment:
Checking for such assertions early  makes sense (as opposed to my earlier suggestion about merging with the for loop)

A trade-off  for more conciseness at the cost of slight less verbose messaging. Feel free to pass on to the next reviewer in case we disagree. 

```
  if (llvm::any_of(rankedTypes, [&](RankedTensorType type) {
        return rank != type.getRank();
      }))
    return emitOptionalError(location, "Expected input types to have same ranks");
```

Reply:
I think printing the ranks is helpful for fault localization